File: tools/generate_clash_cards.dart
// ignore_for_file: avoid_print

import 'dart:convert';
import 'dart:io';

void main() {
  final root = Directory.current.path.contains('eternalxi_front')
      ? Directory('..')
      : Directory.current;
  final playersFile = File('${root.path}/.local/tools/clash_players.tsv');
  final stFile = File('${root.path}/.local/tools/clash_st.tsv');
  final outFile = File('${root.path}/eternalxi_front/assets/data/clash/cards.json');
  final manifestFile =
      File('${root.path}/eternalxi_front/assets/data/clash/cards_manifest.json');

  final players = _loadPlayers(playersFile);
  final stByPlayer = _loadSuperTechniques(stFile);
  final cards = <Map<String, dynamic>>[];
  final teamCounters = <String, int>{};

  for (final player in players) {
    final pid = player.id;
    final pos = player.posicion;
    final dorsal = player.dorsal;
    final position = _clashPosition(pos, dorsal);
    final pcode = _posCode(position);
    final teamSlug = _slugify(player.team);
    final seq = (teamCounters[teamSlug] ?? 0) + 1;
    teamCounters[teamSlug] = seq;
    final style = _styleMap[player.estilo]!;
    final stList = stByPlayer[pid] ?? const [];

    for (final rarity in ['n', 'r', 'sr']) {
      final cardId = '$teamSlug-$rarity-$pcode-${seq.toString().padLeft(3, '0')}';
      final stats = _buildStats(position, player.valoracion, rarity);
      cards.add({
        'id': cardId,
        'playerId': pid,
        'name': player.name,
        'team': player.team,
        'dorsal': dorsal,
        'teamId': player.teamId,
        'rarity': rarity,
        'level': 1,
        'style': style,
        'position': position,
        'basicPortraitPath': 'network',
        'stats': stats,
        'superTechniques':
            _buildStEntries(stList, rarity, cardId),
      });
    }
  }

  final payload = {'schemaVersion': 1, 'cards': cards};
  final raw = const JsonEncoder.withIndent('  ').convert(payload);
  outFile.parent.createSync(recursive: true);
  outFile.writeAsStringSync(raw);
  final sizeBytes = outFile.lengthSync();

  final eternalXiN = cards
      .where((c) => c['team'] == 'Eternal XI' && c['rarity'] == 'n')
      .length;

  manifestFile.writeAsStringSync(
    const JsonEncoder.withIndent('  ').convert({
      'cardsVersion': 4,
      'cardsUrl': 'https://api.eternalxi.com/api/v1/assets/clash/cards.json',
      'cardsBytes': sizeBytes,
      'playerCount': players.length,
      'cardCount': cards.length,
      'portraitsBaseUrl':
          'https://api.eternalxi.com/api/v1/assets/players',
      'portraitsBytesEstimate': players.length * 1650000,
      'portraitsVersion': 1,
    }),
  );

  print('Generated ${cards.length} cards for ${players.length} players');
  print('Eternal XI N cards: $eternalXiN');
  print('Output: ${outFile.path} (${(sizeBytes / 1024 / 1024).toStringAsFixed(2)} MB)');
}

const _styleMap = {
  'PICARO': 'picaro',
  'PRECISO': 'preciso',
  'POTENTE': 'potente',
  'VALIENTE': 'valiente',
  'AGIL': 'agil',
};

const _stTypeMap = {
  'PARADA': 'save',
  'DEFENSA': 'defense',
  'REGATE': 'dribble',
  'TIRO': 'shot',
};

const _posProfile = {
  'goalkeeper': {
    'save': [30, 14],
    'defense': [16, 8],
    'pass': [12, 6],
    'dribble': [6, 4],
    'shot': [4, 2],
    'techniquePoints': [16, 8],
    'stamina': [100, 8],
  },
  'centreBack': {
    'save': [4, 2],
    'defense': [31, 12],
    'pass': [13, 6],
    'dribble': [8, 4],
    'shot': [7, 3],
    'techniquePoints': [16, 8],
    'stamina': [102, 8],
  },
  'fullBack': {
    'save': [4, 2],
    'defense': [25, 11],
    'pass': [16, 7],
    'dribble': [13, 6],
    'shot': [9, 4],
    'techniquePoints': [16, 8],
    'stamina': [103, 8],
  },
  'defensiveMidfielder': {
    'save': [3, 1],
    'defense': [21, 9],
    'pass': [22, 10],
    'dribble': [15, 7],
    'shot': [12, 5],
    'techniquePoints': [18, 8],
    'stamina': [104, 8],
  },
  'attackingMidfielder': {
    'save': [3, 1],
    'defense': [13, 7],
    'pass': [23, 10],
    'dribble': [20, 9],
    'shot': [17, 7],
    'techniquePoints': [18, 8],
    'stamina': [104, 8],
  },
  'winger': {
    'save': [3, 1],
    'defense': [11, 5],
    'pass': [16, 7],
    'dribble': [24, 11],
    'shot': [22, 10],
    'techniquePoints': [18, 8],
    'stamina': [106, 10],
  },
  'striker': {
    'save': [3, 1],
    'defense': [12, 6],
    'pass': [12, 5],
    'dribble': [18, 8],
    'shot': [28, 13],
    'techniquePoints': [18, 8],
    'stamina': [105, 10],
  },
};

const _rarityMult = {'n': 1.0, 'r': 1.08, 'sr': 1.18};
const _rarityStCount = {'n': 2, 'r': 3, 'sr': 4};

class _PlayerRow {
  _PlayerRow({
    required this.id,
    required this.teamId,
    required this.team,
    required this.name,
    required this.pila,
    required this.dorsal,
    required this.posicion,
    required this.estilo,
    required this.valoracion,
  });

  final int id;
  final int teamId;
  final String team;
  final String name;
  final String pila;
  final int dorsal;
  final String posicion;
  final String estilo;
  final int valoracion;
}

class _StRow {
  _StRow({
    required this.orden,
    required this.nombre,
    required this.potencia,
    required this.tipo,
    required this.estilo,
  });

  final int orden;
  final String nombre;
  final int potencia;
  final String tipo;
  final String estilo;
}

List<_PlayerRow> _loadPlayers(File file) {
  return file
      .readAsLinesSync()
      .where((l) => l.trim().isNotEmpty)
      .map((line) {
        final p = line.split('\t');
        return _PlayerRow(
          id: int.parse(p[0]),
          teamId: int.parse(p[1]),
          team: p[2],
          name: p[3],
          pila: p[4],
          dorsal: int.parse(p[5]),
          posicion: p[6],
          estilo: p[7],
          valoracion: double.parse(p[8]).round(),
        );
      })
      .toList();
}

Map<int, List<_StRow>> _loadSuperTechniques(File file) {
  final map = <int, List<_StRow>>{};
  for (final line in file.readAsLinesSync()) {
    if (line.trim().isEmpty) continue;
    final p = line.split('\t');
    final id = int.parse(p[0]);
    map.putIfAbsent(id, () => []).add(
          _StRow(
            orden: int.parse(p[2]),
            nombre: p[7],
            potencia: int.parse(p[4]),
            tipo: p[5],
            estilo: p[6],
          ),
        );
  }
  // Desbloqueo: de la más débil a la más fuerte (potencia ASC).
  for (final list in map.values) {
    list.sort((a, b) {
      final byPower = a.potencia.compareTo(b.potencia);
      if (byPower != 0) return byPower;
      return a.orden.compareTo(b.orden);
    });
  }
  return map;
}

String _slugify(String text) {
  final normalized = text
      .toLowerCase()
      .replaceAll(RegExp(r'[^a-z0-9]+'), '-')
      .replaceAll(RegExp(r'^-|-$'), '');
  return normalized.isEmpty ? 'player' : normalized;
}

String _clashPosition(String pos, int dorsal) {
  switch (pos.toUpperCase()) {
    case 'POR':
      return 'goalkeeper';
    case 'LAT':
    case 'LATERAL':
      return 'fullBack';
    case 'DFC':
    case 'CENTRAL':
    case 'DEFENSA CENTRAL':
      return 'centreBack';
    case 'MCD':
    case 'MEDIOCENTRO ATRASADO':
      return 'defensiveMidfielder';
    case 'MCO':
    case 'MEDIOCENTRO OFENSIVO':
      return 'attackingMidfielder';
    case 'EXT':
    case 'EXTREMO':
      return 'winger';
    case 'DEL':
    case 'DC':
    case 'DELANTERO':
    case 'DELANTERO CENTRO':
      return 'striker';
    // Legado grueso (por si queda alguna fila sin migrar).
    case 'DEF':
      return dorsal.isEven ? 'fullBack' : 'centreBack';
    case 'MED':
      return dorsal <= 8 ? 'defensiveMidfielder' : 'attackingMidfielder';
    default:
      return dorsal.isEven ? 'winger' : 'striker';
  }
}

String _posCode(String position) => switch (position) {
      'goalkeeper' => 'gk',
      'centreBack' => 'cb',
      'fullBack' => 'fb',
      'defensiveMidfielder' => 'dm',
      'attackingMidfielder' => 'am',
      'winger' => 'wg',
      _ => 'st',
    };

Map<String, int> _buildStats(String pos, int valoracion, String rarity) {
  final profile = _posProfile[pos]!;
  final scale = ((valoracion - 70) / 22.0).clamp(0.0, 1.0);
  final mult = _rarityMult[rarity]!;
  return {
    for (final entry in profile.entries)
      entry.key: ((entry.value[0] + entry.value[1] * scale) * mult)
          .round()
          .clamp(1, 999),
  };
}

int _ptCost(int power) {
  if (power >= 70) return 12;
  if (power >= 55) return 10;
  if (power >= 40) return 9;
  return 8;
}

List<Map<String, dynamic>> _buildStEntries(
  List<_StRow> stList,
  String rarity,
  String cardId,
) {
  final count = _rarityStCount[rarity]!;
  final entries = <Map<String, dynamic>>[];
  for (var i = 0; i < count && i < stList.length; i++) {
    final st = stList[i];
    entries.add({
      'id': '$cardId-st${i + 1}',
      'name': st.nombre,
      'description': st.nombre,
      'type': _stTypeMap[st.tipo],
      'style': _styleMap[st.estilo],
      'basePower': st.potencia,
      'ptCost': _ptCost(st.potencia),
      'level': 'normal',
    });
  }
  return entries;
}

File: tools/generate_clash_cards.py
#!/usr/bin/env python3
"""Genera cards.json Clash (N/R/SR) desde export TSV de MySQL Eterno Campeon."""
from __future__ import annotations

import json
import math
import re
import unicodedata
from collections import defaultdict
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
PLAYERS_TSV = ROOT / ".local" / "tools" / "clash_players.tsv"
ST_TSV = ROOT / ".local" / "tools" / "clash_st.tsv"
OUT_JSON = ROOT / "eternalxi_front" / "assets" / "data" / "clash" / "cards.json"
MANIFEST_JSON = ROOT / "eternalxi_front" / "assets" / "data" / "clash" / "cards_manifest.json"

STYLE_MAP = {
    "PICARO": "picaro",
    "PRECISO": "preciso",
    "POTENTE": "potente",
    "VALIENTE": "valiente",
    "AGIL": "agil",
}

ST_TYPE_MAP = {
    "PARADA": "save",
    "DEFENSA": "defense",
    "REGATE": "dribble",
    "TIRO": "shot",
}

# Perfiles finos por subposición Clash (base, spread) a valoración 70→92.
POS_PROFILE = {
    "goalkeeper": {
        "save": (30, 14),
        "defense": (16, 8),
        "pass": (12, 6),
        "dribble": (6, 4),
        "shot": (4, 2),
        "techniquePoints": (16, 8),
        "stamina": (100, 8),
    },
    "centreBack": {
        "save": (4, 2),
        "defense": (31, 12),
        "pass": (13, 6),
        "dribble": (8, 4),
        "shot": (7, 3),
        "techniquePoints": (16, 8),
        "stamina": (102, 8),
    },
    "fullBack": {
        "save": (4, 2),
        "defense": (25, 11),
        "pass": (16, 7),
        "dribble": (13, 6),
        "shot": (9, 4),
        "techniquePoints": (16, 8),
        "stamina": (103, 8),
    },
    "defensiveMidfielder": {
        "save": (3, 1),
        "defense": (21, 9),
        "pass": (22, 10),
        "dribble": (15, 7),
        "shot": (12, 5),
        "techniquePoints": (18, 8),
        "stamina": (104, 8),
    },
    "attackingMidfielder": {
        "save": (3, 1),
        "defense": (13, 7),
        "pass": (23, 10),
        "dribble": (20, 9),
        "shot": (17, 7),
        "techniquePoints": (18, 8),
        "stamina": (104, 8),
    },
    "winger": {
        "save": (3, 1),
        "defense": (11, 5),
        "pass": (16, 7),
        "dribble": (24, 11),
        "shot": (22, 10),
        "techniquePoints": (18, 8),
        "stamina": (106, 10),
    },
    "striker": {
        "save": (3, 1),
        "defense": (12, 6),
        "pass": (12, 5),
        "dribble": (18, 8),
        "shot": (28, 13),
        "techniquePoints": (18, 8),
        "stamina": (105, 10),
    },
}

RARITY_MULT = {"n": 1.0, "r": 1.08, "sr": 1.18}
RARITY_ST_COUNT = {"n": 1, "r": 2, "sr": 3}
PT_COST_BY_POWER = [(70, 12), (55, 10), (40, 9), (0, 8)]


def slugify(text: str) -> str:
    text = unicodedata.normalize("NFKD", text)
    text = text.encode("ascii", "ignore").decode("ascii")
    text = re.sub(r"[^a-zA-Z0-9]+", "-", text).strip("-").lower()
    return text or "player"


def clash_position(pos: str, dorsal: int) -> str:
    if pos == "POR":
        return "goalkeeper"
    if pos == "DEF":
        return "fullBack" if dorsal % 2 == 0 else "centreBack"
    if pos == "MED":
        return "defensiveMidfielder" if dorsal <= 8 else "attackingMidfielder"
    return "winger" if dorsal % 2 == 0 else "striker"


def pos_code(position: str) -> str:
    return {
        "goalkeeper": "gk",
        "centreBack": "cb",
        "fullBack": "fb",
        "defensiveMidfielder": "dm",
        "attackingMidfielder": "am",
        "winger": "wg",
        "striker": "st",
    }[position]


def stat_value(base: float, spread: float, scale: float, mult: float) -> int:
    return max(1, int(round((base + spread * scale) * mult)))


def build_stats(clash_pos: str, valoracion: int, rarity: str) -> dict:
    profile = POS_PROFILE[clash_pos]
    scale = max(0.0, min(1.0, (valoracion - 70) / 22.0))
    mult = RARITY_MULT[rarity]
    return {
        key: stat_value(base, spread, scale, mult)
        for key, (base, spread) in profile.items()
    }


def pt_cost(power: int) -> int:
    for threshold, cost in PT_COST_BY_POWER:
        if power >= threshold:
            return cost
    return 8


def load_players() -> list[dict]:
    rows = []
    for line in PLAYERS_TSV.read_text(encoding="utf-8").splitlines():
        if not line.strip():
            continue
        parts = line.split("\t")
        if len(parts) < 10:
            continue
        rows.append(
            {
                "id": int(parts[0]),
                "teamId": int(parts[1]),
                "team": parts[2],
                "name": parts[3],
                "pila": parts[4],
                "dorsal": int(parts[5]),
                "posicion": parts[6],
                "estilo": parts[7],
                "valoracion": int(float(parts[8])),
                "foto": parts[9] if len(parts) > 9 else "",
            }
        )
    return rows


def load_supertechniques() -> dict[int, list[dict]]:
    by_player: dict[int, list[dict]] = defaultdict(list)
    for line in ST_TSV.read_text(encoding="utf-8").splitlines():
        if not line.strip():
            continue
        parts = line.split("\t")
        if len(parts) < 8:
            continue
        player_id = int(parts[0])
        by_player[player_id].append(
            {
                "orden": int(parts[2]),
                "clave": parts[3],
                "potencia": int(parts[4]),
                "tipo": parts[5],
                "estilo": parts[6],
                "nombre": parts[7],
            }
        )
    for st_list in by_player.values():
        st_list.sort(key=lambda x: x["orden"])
    return by_player


def build_st_entries(player_id: int, st_list: list[dict], rarity: str, card_id: str) -> list[dict]:
    count = RARITY_ST_COUNT[rarity]
    entries = []
    for idx, st in enumerate(st_list[:count], start=1):
        power = st["potencia"]
        entries.append(
            {
                "id": f"{card_id}-st{idx}",
                "name": st["nombre"],
                "description": st["nombre"],
                "type": ST_TYPE_MAP[st["tipo"]],
                "style": STYLE_MAP[st["estilo"]],
                "basePower": power,
                "ptCost": pt_cost(power),
                "level": "normal",
            }
        )
    return entries


def main() -> None:
    players = load_players()
    st_by_player = load_supertechniques()
    cards = []

    team_counters: dict[str, int] = defaultdict(int)

    for player in players:
        pid = player["id"]
        pos = player["posicion"]
        dorsal = player["dorsal"]
        position = clash_position(pos, dorsal)
        pcode = pos_code(position)
        team_slug = slugify(player["team"])
        team_counters[team_slug] += 1
        seq = team_counters[team_slug]

        style = STYLE_MAP[player["estilo"]]
        st_list = st_by_player.get(pid, [])

        for rarity in ("n", "r", "sr"):
            card_id = f"{team_slug}-{rarity}-{pcode}-{seq:03d}"
            stats = build_stats(position, player["valoracion"], rarity)
            card = {
                "id": card_id,
                "playerId": pid,
                "name": player["name"],
                "team": player["team"],
                "dorsal": dorsal,
                "teamId": player["teamId"],
                "rarity": rarity,
                "level": 1,
                "style": style,
                "position": position,
                "basicPortraitPath": "network",
                "stats": stats,
                "superTechniques": build_st_entries(pid, st_list, rarity, card_id),
            }
            cards.append(card)

    payload = {"schemaVersion": 1, "cards": cards}
    OUT_JSON.parent.mkdir(parents=True, exist_ok=True)
    raw = json.dumps(payload, ensure_ascii=False, indent=2)
    OUT_JSON.write_text(raw, encoding="utf-8")

    size_bytes = OUT_JSON.stat().st_size
    manifest = {
        "schemaVersion": 1,
        "cardsVersion": 1,
        "cardsUrl": "https://api.eternalxi.com/api/v1/assets/clash/cards.json",
        "cardsBytes": size_bytes,
        "cardsSha256": None,
        "playerCount": len(players),
        "cardCount": len(cards),
        "portraitsBaseUrl": "https://api.eternalxi.com/api/v1/assets/players",
        "portraitsBytesEstimate": len(players) * 1650000,
        "portraitsVersion": 1,
    }
    MANIFEST_JSON.write_text(json.dumps(manifest, indent=2), encoding="utf-8")

    eternal_xi_n = sum(
        1 for c in cards if c["team"] == "Eternal XI" and c["rarity"] == "n"
    )
    print(f"Generated {len(cards)} cards for {len(players)} players")
    print(f"Eternal XI N cards: {eternal_xi_n}")
    print(f"Output: {OUT_JSON} ({size_bytes / 1024 / 1024:.2f} MB)")


if __name__ == "__main__":
    main()
